compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/test/iree_gpu_ops.mlir

```diff
@@ -92,3 +92,143 @@ func.func @tensor_barrier_multiple_inputs(%input: tensor<?xf16>) -> (tensor<?xf1
+
+// -----
+
+// Test coalesced_gather_dma within 2-level nested scf.forall
+func.func @coalesced_gather_dma_in_forall(%indices: tensor<16x32xindex>, %source: tensor<2048x64xf32>, %dest: tensor<128x16xf32>) -> tensor<128x16xf32> {
+  // Outer forall: warp level parallelism
+  %result = scf.forall (%wg_i, %wg_j) in (16, 1) shared_outs(%wg_out = %dest) -> (tensor<128x16xf32>) {
+    %c8 = arith.constant 8 : index
+    %wg_offset = arith.muli %wg_i, %c8 : index
+
+    %indices_wg_slice = tensor.extract_slice %indices[%wg_offset, 0] [1, 32] [1, 1]
+      : tensor<16x32xindex> to tensor<1x32xindex>
+
+    %dest_wg_slice = tensor.extract_slice %wg_out[%wg_offset, 0] [8, 16] [1, 1]
+      : tensor<128x16xf32> to tensor<8x16xf32>
+
+    // Inner forall: thread level parallelism
+    %inner_result = scf.forall (%sg_i, %sg_j) in (32, 1) shared_outs(%sg_out = %dest_wg_slice) -> (tensor<8x16xf32>) {
+      scf.forall.in_parallel {
+        iree_gpu.coalesced_gather_dma %indices_wg_slice, %source into %sg_out with tensor<4xf32>
+          : tensor<1x32xindex>, tensor<2048x64xf32>, tensor<8x16xf32> -> tensor<8x16xf32>
+      }
+    } {mapping = [#gpu.thread<linear_dim_1>, #gpu.thread<linear_dim_0>]}
+
+    scf.forall.in_parallel {
+      tensor.parallel_insert_slice %inner_result into %wg_out[%wg_offset, 0] [8, 16] [1, 1]
+        : tensor<8x16xf32> into tensor<128x16xf32>
+    }
+  } {mapping = [#gpu.warp<linear_dim_1>, #gpu.warp<linear_dim_0>]}
+  return %result : tensor<128x16xf32>
+}
+
+// CHECK-LABEL: func @coalesced_gather_dma_in_forall
+//  CHECK-SAME:   %[[INDICES:[A-Za-z0-9]+]]: tensor<16x32xindex>
+//  CHECK-SAME:   %[[SOURCE:[A-Za-z0-9]+]]: tensor<2048x64xf32>
+//  CHECK-SAME:   %[[DEST:[A-Za-z0-9]+]]: tensor<128x16xf32>
+//       CHECK:   %[[RESULT:.+]] = scf.forall (%[[WG_I:.+]], %[[WG_J:.+]]) in (16, 1) shared_outs(%[[WG_OUT:.+]] = %[[DEST]])
+//       CHECK:     %[[C8:.+]] = arith.constant 8 : index
+//       CHECK:     %[[WG_OFFSET:.+]] = arith.muli %[[WG_I]], %[[C8]]
+//       CHECK:     %[[INDICES_WG_SLICE:.+]] = tensor.extract_slice %[[INDICES]][%[[WG_OFFSET]], 0] [1, 32] [1, 1] : tensor<16x32xindex> to tensor<1x32xindex>
+//       CHECK:     %[[DEST_WG_SLICE:.+]] = tensor.extract_slice %[[WG_OUT]][%[[WG_OFFSET]], 0] [8, 16] [1, 1] : tensor<128x16xf32> to tensor<8x16xf32>
+//       CHECK:     %[[INNER_RESULT:.+]] = scf.forall (%[[SG_I:.+]]) in (32) shared_outs(%[[SG_OUT:.+]] = %[[DEST_WG_SLICE]])
+//       CHECK:       scf.forall.in_parallel
+//       CHECK:         iree_gpu.coalesced_gather_dma %[[INDICES_WG_SLICE]], %[[SOURCE]] into %[[SG_OUT]] with tensor<4xf32> : tensor<1x32xindex>, tensor<2048x64xf32>, tensor<8x16xf32> -> tensor<8x16xf32>
+//       CHECK:     }
+//       CHECK:     scf.forall.in_parallel
+//       CHECK:       tensor.parallel_insert_slice %[[INNER_RESULT]] into %[[WG_OUT]][%[[WG_OFFSET]], 0] [8, 16] [1, 1] : tensor<8x16xf32> into tensor<128x16xf32>
```

Comment:
This test is way too involved for a parse/print test (iree_gpu_ops.mlir). The idea behind the test looks good, but save this test for somewhere else.

One option is to put it in the op description if you want.

Reply:
If you want to test this scenario, a parser/printer/verifier test is the wrong place. This should be a part of a lowering test of some kind to make sure the compiler handles it properly. Really parser/printer/verifier tests are only worth it for custom parser/printer implementations and should be quite sparse and as simple as possible. Pulling in IR from other dialects is generally bad form for a test like this but since the verifier requires `scf.forall.in_parallel`, the single level loop nest is OK, but a double level loop nest is overkill.